### DISCUSSION TITLE
Birefringence and computations on GPU vs CPU

### DISCUSSION THREAD

**Discussion by AndiGaw** [AaHaHaa/MMTools#5] [Q&A] [answered]
Hello,

Thank you for sharing your code. It is very useful and I am going to cite it once I am done with my paper(s). 

I use your code to solve MMGNLSE for a birefringent fiber and I have a question. Namely, If I understand correctly, when considering birefringence and coupling between different spacial modes, your polarization code (sim.scalar=false) does not longer give correct results and cannot be used to calculate SK. For that reason, I run scalar simulations and provide a separate SK tensor for calSRSK.m file along with the SR tensor (this is since SK =/= SR). Further, for sanity check, I compare calculations done on GPU and CPU but I get different results. Since I save the SK/SR used in the computation, the clear difference between the GPU/CPU calculations is the number of SK/SR elements. I think it is linked to the summation S13 (from the supplementary materials). Is the summation S13 rigorous, or is it some sort of a physical approximation? As far as I understood, it is only used for the GPU computations, right? 

Thanks in advance and best regards,
Andrzej

**Comment by AaHaHaa** [comment]
Hi, first thanks for using my code.

There was a bug in multimode polarization computation, but I fixed it last year. If you run into this problem, please re-download the code. I'm afraid you're using the old version. It can run correctly now.

I've checked very carefully the new code last year, so I think the current one should solve the polarization mode correctly. Could you please confirm with me that you're using the latest code?

If you're using the latest code, then please send me the script. I can test it myself for you.

About the old polarization bug, the problem lies in cuda's summation. I forgot to put int(midx & 1) in if-else statements, and the wrong "if (midx & 1)" makes no sense since "midx & 1" is a binary number. I added int(....) to fix it.

Since the old bug lied in cuda, the CPU computation should always be the correct one. (It should be......I cannot check it now; everything is updated to the new correct one.)

**Comment by AaHaHaa** [comment]
I am thinking about your statement:
"Namely, If I understand correctly, when considering birefringence and coupling between different spatial modes, your polarization code (sim.scalar=false) does not longer give correct results and cannot be used to calculate SK."

My code should be able to solve multimode polarization cases, so I don't understand your statements. The calculation of SK values is independent of birefringence since they're overlap integrals, only related to functions of x and y.

Due to your inconsistency between CPU and GPU results, as I mentioned previously, I kinda, or strongly, speculate that you might be using my old-version buggy code.

**Reply by AaHaHaa** [reply]
Another point is that
In GPU cuda, I've optimzied the computation with reducing computations with symmetry from real and imaginary parts (see the details in the supplement of my JOSAB multimode-gain paper). This, however, isn't implemented in CPU. CPU uses MATLAB complex number computations only. I kinda don't assume that people will use CPU for multimode scenarios.
That said, you might see inconsistency in "matrix size" in their SRa (or b) and SK arrays, due to implementations of computational reduction in GPU. However, you should be able to find the same numeric values between CPU's and GPU's.

**Comment by AndiGaw** [comment]
Hello, 

Thank you for your swift reply.

Indeed, I have been using the old version of the code, which I downloaded just before the update. 

I am now running a couple of tests to check if the problem is gone. I will let you know asap.

Best regards,
Andrzej

**Reply by AndiGaw** [reply]
Hello,

Indeed, you are right about the correctness of the SK matrices while using the updated polarization code. However, I would still like to full understand how your code deals with polarized modes. 

What I want to do is to calculate interactions (e.g. FWM) between different modes of a birefringent fiber (e.g., LP01(x) and LP01(y) modes, or LP01(x), LP01(y), LP11(xo), and LP11(ye) modes). Following the example of the Random Mode Coupling, I understand that you have 6 modes and load betas and the SR tensor for all of them. Then, you select the modes 1-3 and set _num_modes = length(sim.midx)*2_ (is it necessary?), which generates the input field consisting of 6 fields. However, if you were to select more modes, you would end up with more than 6 modes at the input, while you provide the betas and SRs for 6 modes only. Isn't that a problem?

In my case, I would like to excite 2 modes (e.g. LP01(x) and LP01(y)) while accounting for polarization (for SK calculation). However, the input field need to consists of 4 modes (_num_modes = length(sim.midx)*2_) to avoid the _check_nummodes error_. To get rid of the two "extra" modes, I multiply the input fields by [1, 1, 0, 0]. However, if I do so, I don't observe the desired processes. They appear when I excite all four input fields. I guess, it means I am not _probing_ the SR/SK tensor properly. This leads me to the following questions:

1. Is _num_modes = length(sim.midx)*2_ necessary?
2. If you set _num_modes = length(sim.midx)*2_, what are these "extra" modes? Copies of the selected ones?
3. Is there a way to excite only 2 modes in this case, or you always have to consider these extra ones?

Best regards,
Andrzej

**Reply by AaHaHaa** [reply]
There are a few points I'd like to point out, which are also mentioned in my "readme.pdf":

My code can extend scalar betas to polarized betas if necessary (and in a non-PM fiber of course) by assuming that polarized modes of the same spatial mode see the same beta.

Polarized `num_modes = length(sim.midx)*2` is used only for generating the initial fields; it isn't sent into the GMMNLSE_propagate() function.

We need `num_modes` to create a 6-mode input field (Nt-by-6 matrix) where 6 is composed of 3 spatial modes with 2 polarized modes for each spatial mode. Otherwise, polarized simulation for 3 spatial modes but with only 3-mode field doesn't make sense.

In your case, your problem lies in creating all-zero input fields for the HOMs (it and its polarized mode). The GMMNLSE/UPPE gives all-zero result if the input field of that mode is perfectly zero. You need to excite with `[1,1,1e-3,1e-3]` for example, instead of `[1,1,0,0]`.

If you would like to run multimode simulations with only scalar fields, then set `sim.scalar=false;` with only scalar fields (e.g. `num_modes=length(sim.midx);`). You cannot have `sim.scalar=false;` and initial fields excluding polarized fields simultaneously.

I hope these answer your question. I didn't answer yours one by one because I realized that they're all related to the misunderstanding of the polarized-mode operation of the code.

**Reply by AaHaHaa** [reply]
By the way, see my readme.pdf:
The polarized fields is arranged as [A1x,A1y, A2x,A2y...] for example. If you want to study the polarized multimode case with excited x-polarized multimode fields, you should multiply with [1,1e-3,1,1e-3] so that there are weakly-excited y-polarized components with mainly x-polarized inputs.
Since you mentioned `[1,1,0,0]', I'd like to notify you of this in case you wrongly thought it were [A1x,A2x,A1y,A2y].

**Reply by AndiGaw** [reply]
Thank you for your reply.

_My code can extend scalar betas to polarized betas if necessary (and in a non-PM fiber of course) by assuming that polarized modes of the same spatial mode see the same beta._ But can it be used to calculate a scenario where the polarized modes of the same spatial modes see different betas? 

_If you would like to run multimode simulations with only scalar fields, then set sim.scalar=false; with only scalar fields (e.g. num_modes=length(sim.midx);). You cannot have sim.scalar=false; and initial fields excluding polarized fields simultaneously._ Sure, but doesn't setting `sim.scalar=false;` make SK = SR, which is not the case for polarized fields?

**Reply by AaHaHaa** [reply]
Of course it can calculate birefringent multimode cases, haha.
The single-mode birefringent case is in Examples/Gain-rate-equation model/birefringent_YDFA.m. You can easily extend it to multimode cases. The idea is to make `betas=[beta_1x,beta_1y,beta_2x,beta_2y......]` for example.

`sim.scalar=true` makes SK=2/3*SR for circularly polarized scalar field (with `sim.ellipticity=1`) or SK=SR for linearly polarized scalar field (with `sim.ellipticity=0`). `sim.scalar=false` computes SR and SK separately with their own overlap integrals. If you're interested, see the function calc_SRSK.m in the GMMNLSE_algorithm/ folder. The polarized SRSK computation is quite complicated.

**Reply by AaHaHaa** [reply]
In calc_SRSK.m, I separately compute them for linear and circular polarizations because these two polarizations have symmetry and thus the number of plmn overlap-integral summations is reduced with some being zero. For example, those with inner_product(F_ix,F_jy)=0 whatever spatial modes i and j are.

**Reply by AndiGaw** [reply]
Thanks. I think the problem I had was indeed due to the previously (bugged) version of the code. I will try the birefringent_YDFA.m with multiple modes and change the size of SR accordingly.

I am quite familiar with the calc_SRSK.m function as, while working with the previous version of the code, I tried to solve some problems by providing the load_default_GMMNLSE_propagate with both SR and SK :).

**Reply by AaHaHaa** [reply]
Good luck. If you have more questions, just feel free to ask me. I'll answer them ASAP if I can. I'm also happy that there are people out there using my code, which is the main purpose of me sharing this code.

(I'm thinking about how to share it more widely, but I have no good solution now. Currently, it seems that only those reading my JOSAB paper know it. However, it can run simulations with situations beyond multimode gain. Even single-mode co- and counter-pumping are very useful for many people out there designing fiber amplifiers. They're much better than Gaussian gain that doesn't give pump info.)

**Reply by AndiGaw** [reply]
Thanks! 

Actually, I still have one question. Namely, to better understand the code, I am trying to get the same results using scalar and polarized versions of your code. To be more specific, I consider two modes and for _polarized version_ of the code provide SR = 1/Aeff. To repeat it with the _scalar version_ of the code, I use the same betas but also provide SR and SK (both 2x2x2x2) with elements obtained from SR(1,1,1,1) using 1/3, 2/3, and 1 factors from your _readme_. To make sure that both versions of the code use the same SR/SK I record them after the computation. Although I am using the same SR and SK, I am getting different results. What I also realized is that I can repeat the results obtained with _polarized version_ using the _scalar version_ only when running the computations on CPU instead of GPU. For the _polarized version_ both GPU and CPU computations give the same results. I believe it is linked to the stepping_MPA function where selecting the scalar/polarized cases impacts the GPU computations but not CPU ones. Would you recommend dropping the if/else statement (in stepping_MPA) and always using the part of the code for _polarized fields_?

Concerning sharing of your code, I think that with time more people will use it for research and cite it :).

**Reply by AaHaHaa** [reply]
Let me run a random simulation to check this. However, this sounds so weird because scalar model is easy to implement and I kinda doubt that it is buggy. Intuitively, scalar implementation cannot be buggier than the polarized one.

**Reply by AaHaHaa** [reply]
I just checked my code. Wow! I found a new bug related to spontaneous Raman scattering. Please update the code. In fact, I modified only all the MPA stepping functions. However, it shouldn't affect what you were seeing because it's a CUDA error that fails MATLAB right away.

I uploaded new examples in Examples/Tutorial/MM soliton/.
I didn't see the inconsistency you have. I tried to run with scalar and polarized cases, and CPU and GPU cases. They all gave me the same results.

When I created the examples, I realized the necessity of carefulness about the betas, or possibly SRSK. For example, I initially forgot that they were loaded from betas.mat and put in typical value for SMF and created weird results. They all work well now. Perhaps try to modify them after load_default_GMMNLSE_propagate().

**Reply by AaHaHaa** [reply]
> I consider two modes and for polarized version of the code provide SR = 1/Aeff. To repeat it with the scalar version of the code, I use the same betas but also provide SR and SK (both 2x2x2x2) with elements obtained from SR(1,1,1,1) using 1/3, 2/3, and 1 factors from your readme. 

I kinda guess that your problem results from incorrect betas or SRSK you sent into. My examples work well. If it's the bug of your script, then I cannot debug it on my side now. Besides, if you calculate your own assuming circular polarization but sim.eiilpticity=0 (linear polarization), then there's inconsistency between those computed automatically by my code and yours. Make sure sim.ellipticity is the same as the polarization you have in mind.

> Would you recommend dropping the if/else statement (in stepping_MPA) and always using the part of the code for polarized fields?

I don't recommend dropping it. The reason I have two implementations is because scalar simulation has less computation and is faster. In principle, one polarized implementation is enough by automatically extending all scalar cases to polarized cases (by adding zeros to the other polarization of the fields) but this induces more computation. Since there is a strong need for scalar simulations for idea demonstration and so on, having separately one faster (than polarized one) scalar implementation is useful. Besides, scalar implementation uses RK4IP while multimode one uses MPA. RK4IP has less numerical truncation loss, so it makes scalar simulations run much faster. Sadly, there is no fast RK4IP generalization to multimode; I mean the one with RK4IP but also in combination with MPA. A simple RK4IP takes into account betas naively, so typical multimode cases with huge beta difference between modes make RK4IP so slow. I have tried to implement RK4IP within MPA before to see if I could combine them (and made MPA have low truncation loss), but I stopped working on it and turned to focus on making MPA with adaptive-step-size control. My remaining fight for reducing truncation loss of MPA is to successfully incorporate Adams-Moulton method, which reduces the loss by a factor of M, MPA's parallelization; better than nothing...... Hope someday people can find a way to integrate MPA and RK4IP, further speeding up multimode simulations. My JOSAB paper didn't emphasize this, but "adaptive-step" MPA is also new to the world. It's probably the fastest and the most convenient multimode numerical scheme now.

**Reply by AndiGaw** [reply]
>I just checked my code. Wow! I found a new bug related to spontaneous Raman scattering. Please update the code. In fact, I modified only all the MPA stepping functions. However, it shouldn't affect what you were seeing because it's a CUDA error that fails MATLAB right away.

Thanks, I updated the files. But as you said, they are not linked to my issue.

> I kinda guess that your problem results from incorrect betas or SRSK you sent into. My examples work well. If it's the bug of your script, then I cannot debug it on my side now. Besides, if you calculate your own assuming circular polarization but sim.eiilpticity=0 (linear polarization), then there's inconsistency between those computed automatically by my code and yours. Make sure sim.ellipticity is the same as the polarization you have in mind.

I will check the calc_SRSK function. I am not worried about sim.eiilpticity, as I basically removed the SK factor and load SK along with SR. So the only place it can fail is when it is send to GPU (as CPU computations give correct results).

> I don't recommend dropping it. The reason I have two implementations is because scalar simulation has less computation and is faster. 

Indeed, the scalar simulations are ~3 times faster for me, so if I have to run 10000 cases it is a game changer.

**Reply by AaHaHaa** [reply]
> I will check the calc_SRSK function. I am not worried about sim.eiilpticity, as I basically removed the SK factor and load SK along with SR. So the only place it can fail is when it is send to GPU (as CPU computations give correct results).

If you can make your test script simple, such as modifying from my MM_soliton or at least using my fiber files (rather than yours, and then I need to download many MB of your beta and SR), you can post it here (copy-paste the entire script I guess) and then I can check it for you quickly. This doesn't sound difficult and won't take a lot of time to me.

In principle, you don't need to modify calc_SRSK(). If you modify it, there is a chance of generating problems and bugs. Even for me understanding everything and am the code author, modifying that part is quite scary. The `beginning_nonzero` and `ending_nonzero` are to control the number of summation for each (p,l) term in overlap integral computation. Be careful not to mess this up, or be careful that you take this into account if your change can disrupt it.

My GPU computation relies on `beginning_nonzero` and `ending_nonzero`. You need to understand these two, if your change can disturb them, to make sure GPU computation is correct. My CUDA is parallelized with (p,l). For example, if you have 2 modes, CUDA has 4 parallelizations. Each (p,l) set has its own overlap-integral summation (this_p,this_l,m,n) over (m,n). These two parameters remember where to sum over in SRSK values, which are linearized to column vectors after removing zeros due to symmetry. Without these two parameters, each (p,l) set will sum over all SRSK's and waste time on going over those that don't belong each (p,l).

I somehow have a feeling that you modified my calc_SRSK(), haha. I can check it for you too if it's not heavily changed.

**Reply by AndiGaw** [reply]
> I uploaded new examples in Examples/Tutorial/MM soliton/.
I didn't see the inconsistency you have. I tried to run with scalar and polarized cases, and CPU and GPU cases. They all gave me the same results.

Do you mean birefringent_MM_soliton.m file? If I execute it, it works for `sim.scalar = false;` but `sim.scalar = true;` gives the check_nummodes error:

_Error using check_nummodes
The number of spatial modes of betas and SR tensors should be the same._

_Error in GMMNLSE_propagate_with_adaptive (line 197)
check_nummodes(sim, fiber, initial_condition.fields);_

_Error in GMMNLSE_propagate (line 74)
foutput = GMMNLSE_propgation_func(fiber, initial_condition, sim, gain_rate_eqn);_

_Error in birefringent_MM_soliton (line 76)
prop_output = GMMNLSE_propagate(fiber,input_field,sim);_

, or do you compare birefringent_MM_soliton.m + `sim.scalar = false;` with MM_soliton.m + `sim.scalar = true;`? If I compare birefringent_MM_soliton.m + `sim.scalar = false;` with MM_soliton.m + `sim.scalar = true;` the results are qualitatively the same, but if you subtract fields from both simulations there are some small quantitative differences.

**Reply by AndiGaw** [reply]
> If you can make your test script simple, such as modifying from my MM_soliton or at least using my fiber files (rather than yours, and then I need to download many MB of your beta and SR), you can post it here (copy-paste the entire script I guess) and then I can check it for you quickly. This doesn't sound difficult and won't take a lot of time to me.

I am working on it but you are right, I modified calc_SRSK() so I am cleaning it up so that there are as little changes w.r.t. the original file as possible.

> In principle, you don't need to modify calc_SRSK().

That's the part with which I partially agree. My goal is to provide my own SK tensor, so I need to overwrite the SK values from calc_SRSK(), right? Of course I could do it after invoking the calc_SRSK() in the GMMNLSE_propagate_with_adaptive(), but I think it is easier to do it inside the calc_SRSK() were you define your `beginning_nonzero` and `ending_nonzero`. That being said, I traced the `beginning_nonzero` and `ending_nonzero` of the SK tensor calculated with your calc_SRSK() and `sim.scalar = false;` and mine calc_SRSK() and `sim.scalar = true;` and they are the same. The only difference is in the stepping_MPA_nogain_adaptive(), where scalar/polarized GPU simulations are calculated using different codes (line 123 vs 145) and require different input parameters, e.g. the scalar simulations don't need `beginning_nonzero` and `ending_nonzero` of the SK tensor, as (if I understand correctly) they use the SR ones. Since the scalar simulations are faster is it possible to use them along with separate `beginning_nonzero` and `ending_nonzero` of the SK tensor (different than these of SR tensor), or would such modification lead to the code for the polarized calculations anyway?

**Reply by AaHaHaa** [reply]
> , or do you compare birefringent_MM_soliton.m + sim.scalar = false; with MM_soliton.m + sim.scalar = true;? If I compare birefringent_MM_soliton.m + sim.scalar = false; with MM_soliton.m + sim.scalar = true; the results are qualitatively the same, but if you subtract fields from both simulations there are some small quantitative differences.

Haha, I used MM_soliton.m for scalar and birefringent_MM_soliton.m for polarized. I didn't realize that you use a single birefringent_MM_soliton.m for both. I fixed the bug in birefringent_MM_soliton.m. Its scalar situation was wrong because of small stupid mistakes in this script, not the propagate functions. Please update birefringent_MM_soliton.m.
With an updated one, you can see it gives all the same results with all various situations (scalar vs. polarized, GPU vs. CPU).

**Reply by AaHaHaa** [reply]
I think you should use polarized simulations.

The fact that scalar is faster is because it has less summations. If you write your own SK, it doesn't reduce the number of summation terms compared to polarized case. You're making scalar case the same as polarized case now.

I think you should just run with my polarized scheme.

I'm the stubborn person who always implements the best scheme to the code. If there is some smart fast way to run polarized case, I would definitely implement it to the polarized scheme as well, which means that your current trying to get my scalar simulation run like polarized case doesn't gain you any speed improvement. If there is such a method, I would make my polarized case faster as well.

**Reply by AaHaHaa** [reply]
I see that you're using the SK equation in my readme.pdf to implement your SK. Then why not run with my polarized scheme?
I previously thought that you had special need for special SK, but now I think you're simply making my scalar scheme as polarized scheme. Doing this won't make your polarized simulation faster (if you have everything correct).

**Reply by AndiGaw** [reply]
> I see that you're using the SK equation in my readme.pdf to implement your SK. Then why not run with my polarized scheme?
I previously thought that you had special need for special SK, but now I think you're simply making my scalar scheme as polarized scheme. Doing this won't make your polarized sim> I see that you're using the SK equation in my readme.pdf to implement your SK. Then why not run with my polarized scheme?
I previously thought that you had special need for special SK, but now I think you're simply making my scalar scheme as polarized scheme. Doing this won't make your polarized simulation faster (if you have everything correct).

I think I should have stressed my goal at the very beginning to make things simpler. For simplicity, let's consider 2 modes. I want to calculate a scenario with birefringence, where SR(1,1,1,1) =/= SR(2,2,2,2). In your polarization code, you are using a single SR value. If I understood correctly, it assumes that the self-overlaps of mode 1 and mode 2 are equal (SR(1,1,1,1) = SR(2,2,2,2)), which is not always the case. By running scalar simulations, I can provide an SR tensor, where SR(1,1,1,1) =/= SR(2,2,2,2). However, the scalar code assumes SK = SR (or 2/3 SR for circular polarization), while, in reality, SK = 1/3 SR, 2/3 SR, or SR as in your readme.pdf.

The way around, I think is the easiest to implement my case, is to change line 158 of calcSRSK(): 

`SK = mode_info.SRa.*[1 1 2/3 2/3 1/3 1/3];`

to 

`SK = mode_info.SK;`,

which I load along with SR. Of course, I have to make sure the size of SK is correct. My only idea was wheter it can be run with faster scalar scheme, which, as you said is not the case.

**Reply by AaHaHaa** [reply]
So you need to run with cases that have different SR/SK's for different polarization modes but the same spatial mode, is this correct?

**Reply by AndiGaw** [reply]
Yes, for 2 modes scenario. For 4 modes (next step): 2 spatial modes times 2 polarizations.

**Reply by AaHaHaa** [reply]
My suggestion:
Try to run with polarized case, and use your modification: `SK = mode_info.SK;` in line 158 of calc_SRSK().

When calculating beginning_nonzeros and ending_nonzeros, it also runs `refine_scalar/polarized_S`, which removes redundant SR/SK's. This operation is different for scalar and polarized cases. Please redownload my calc_SRSK.m; I've added more comments.

**Reply by AaHaHaa** [reply]
A curious question:
Are you doing research on spatial modes that are different with different polarizations?

This model kinda builds upon LP modes, whose polarization is independent of spatial distribution/mode. We can have LP01 in (x or y) or in (+,-). In the HE basis, polarization is coupled with spatial modes. I don't think my code can deal with HE mode. If you're trying to make it work with HE modes, I think you need to rewrite a lot of things. I don't recommend that you use my code for HE modes and hope that it just works. The math is the same, but the code needs a huge change. There must be a lot of extra work to have a HE version of code; for example, SR and SK are now polarization-dependent with HE modes.

**Reply by AaHaHaa** [reply]
If you would like to run with some very special cases, such as with higher-order HE modes, I would recommend that you take off `refine_scalar/polarized_S` or other commands and let the code sums over all potential M^4 overlap integrals. Forget about speed and optimization. Make it work first. You might need to modify the cuda, taking off `beginning_nonzeros` and `ending_nonzeros` to remove these optimization processes.
CUDA's summing over all overlap integrals (though some might be redundant) is still faster than CPU computation.

To be honest, it might be better for you to use our old-version code, which has no optimization. It's easier for you to modify it. For example, I think its CUDA has no symmetry-related optimization I mentioned.

**Reply by AndiGaw** [reply]
> My suggestion: Try to run with polarized case, and use your modification: `SK = mode_info.SK;` in line 158 of calc_SRSK().
> 
> When calculating beginning_nonzeros and ending_nonzeros, it also runs `refine_scalar/polarized_S`, which removes redundant SR/SK's. This operation is different for scalar and polarized cases. Please redownload my calc_SRSK.m; I've added more comments.

Thank you. I re-downloaded it. I have one question concerning it. Namely, while trying to provide my SR/SK and build these structs with modified calc_polarized_SRSK function I get the following error:

Error using parallel.gpu.CUDAKernel/feval
Number of input arguments for this kernel is fixed at 13

If I am not mistaken, this is linked to the fact that sim.cuda_SRSK.NumRHSArguments = 22 for polarized case and 13 for scalar simulations. Since, I cannot find sim.cuda_SRSK file I wanted to ask is this error linked to some mistake in calc_SRSK or I have to change something elsewhere in the code?

> A curious question:
Are you doing research on spatial modes that are different with different polarizations?

I am doing research on the x/y polarized LP modes of a birefringent fiber. For the moment, I am not planning to run HE, but thank you for the info :). 

> To be honest, it might be better for you to use our old-version code, which has no optimization.

Do you mean this one: Multimode nonlinear fiber optics: Massively parallel numerical solver, tutorial, and outlook?

**Reply by AndiGaw** [reply]
> Since, I cannot find sim.cuda_SRSK file I wanted to ask is this error linked to some mistake in calc_SRSK or I have to change something elsewhere in the code?

Nvm, I found it in setup_stepping_kernel ;)

**Reply by AaHaHaa** [reply, marked as answer]
> If I am not mistaken, this is linked to the fact that sim.cuda_SRSK.NumRHSArguments = 22 for polarized case and 13 for scalar simulations. Since, I cannot find sim.cuda_SRSK file I wanted to ask is this error linked to some mistake in calc_SRSK or I have to change something elsewhere in the code?

I think you also need to modify my cuda to change the number of input arguments.

> Do you mean this one: Multimode nonlinear fiber optics: Massively parallel numerical solver, tutorial, and outlook?

Not really; that's just the paper discussing what MPA is. You need our old-version code. See my introduction of this package in the first page of Github. If I were you who needs a huge change of code, it would be easier to write one from scratch.

> I am doing research on the x/y polarized LP modes of a birefringent fiber. For the moment, I am not planning to run HE, but thank you for the info :).

How does the polarization significantly affect the LP spatial mode, which is the case that you're thinking about? But anyway, this is just my curiosity. Just want to remind you that if polarization can be decoupled from spatial mode, then you don't need to modify any of my code to run multimode polarization simulations, in case that you forget this.

**Comment by AndiGaw** [comment]
> By the way, see my readme.pdf: The polarized fields is arranged as [A1x,A1y, A2x,A2y...] for example. If you want to study the polarized multimode case with excited x-polarized multimode fields, you should multiply with [1,1e-3,1,1e-3] so that there are weakly-excited y-polarized components with mainly x-polarized inputs. Since you mentioned `[1,1,0,0]', I'd like to notify you of this in case you wrongly thought it were [A1x,A2x,A1y,A2y].

I understood the [mode1+ mode1_ mode 2+ mode2_...] arrangement. I was referring to which modes serve as the input when `num_modes = length(sim.midx)*2` is used. If I use mode1+ mode1_ as my input, I guess, `num_modes = length(sim.midx)*2` treats each of them as a different spatial mode, while, in reality, they are differently polarized modes of the same spatial mode.

**Reply by AaHaHaa** [reply]
> num_modes = length(sim.midx)*2 treats each of them as a different spatial mode, while, in reality, they are differently polarized modes of the same spatial mode.

This sounds confusing to me.

`num_modes=length(sim.midx)*2` isn't sent into GMMNLSE_propagate(). There is no `num_modes` as an input argument of the propagate function. If you would like to create an initial field with polarized field in spatial mode 1 (weakly-excited field in spatial mode 2), then multiply a Gaussian with `[1,1,1e-3,1e-3]` for example if two  (spatial mode 1)'s polarization modes are equally excited without phase difference. Perhaps you were simply rephrasing this; if yes, then we reach consensus.
